divi/src/ChainstateManager.cpp

```diff
@@ -16,11 +16,39 @@ ChainstateManager* instance = nullptr;
 /** Lock for accessing the instance global.  */
 CCriticalSection instanceLock;
 
+class CCoinsViewErrorCatcher : public CCoinsViewBacked
+{
+public:
+  CCoinsViewErrorCatcher(CCoinsView* view) : CCoinsViewBacked(view) {}
+  bool GetCoins(const uint256& txid, CCoins& coins) const override;
+  // Writes do not need special protection, as failure to write is handled by
+  // the caller.
+};
+
+bool CCoinsViewErrorCatcher::GetCoins (const uint256& txid, CCoins& coins) const
+{
+  try {
+    return CCoinsViewBacked::GetCoins(txid, coins);
+  } catch (const std::runtime_error& e) {
+    uiInterface.ThreadSafeMessageBox(translate("Error reading from database, shutting down."), "", CClientUIInterface::MSG_ERROR);
+    LogPrintf("Error reading from database: %s\n", e.what());
+    // Starting the shutdown sequence and returning false to the caller would be
+    // interpreted as 'entry not found' (as opposed to unable to read data), and
+    // could lead to invalid interpration. Just exit immediately, as we can't
+    // continue anyway, and all writes should be atomic.
+    abort();
+  }
+}
+
 } // anonymous namespace
 
-ChainstateManager::ChainstateManager ()
-  : blockMap(mapBlockIndex), activeChain(chainActive),
-    blockTree(*pblocktree), coinsTip(*pcoinsTip)
+ChainstateManager::ChainstateManager (const size_t blockTreeCache, const size_t coinDbCache,
+                                      const bool fMemory, const bool fWipe)
+  : blockMap(new BlockMap ()), activeChain(new CChain ()),
+    blockTree(new CBlockTreeDB (blockTreeCache, fMemory, fWipe)),
+    coinsDbView(new CCoinsViewDB (*blockMap, coinDbCache, fMemory, fWipe)),
+    coinsCatcher(new CCoinsViewErrorCatcher (coinsDbView.get ())),
+    coinsTip(new CCoinsViewCache (coinsCatcher.get ()))
 {
```

Comment:
* activeChain wasn't initialized
* The unique pointers aren't being reset in the destructor, and explicit destruction is preferred.

Reply:
The `activeChain` is initialised, but I've now moved it to its own line to make it easier to see.

The pointers are automatically destructed in reverse order of declaration in the class (which is what it should be).  But I agree it is probably clearer to explicitly `reset` them in the desired order in the destructor, which I've added now to the PR.